Aws cloudwatch source type (#548)
observIQ/bindplane-op: docs/www/integrations/sources/aws-cloudwatch.md

Some prerequisites:

1. Creating AWS Account
2. Creating an IAM user account
   - Required Permissions
     - `logs:GetLogEvents`
     - `logs:DescribeLogGroups`
     - `logs:DescribeLogStreams`
3. Create an access key ID and secret access key

---

**NOTE**

In the getting started guide, it will instruct you to [install](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) upon either the local user or for all users. This is standard for AWS CLI, however in order to run BP Agents AWS Cloudwatch source, users should install the CLI as the root user/ all users.

---

### Credentials

### Credential and Config Files

> AWS Authentication utilizes a user profile specified in a `.aws/credentials` file. Each profiles' credentials should include at minimum

Reviewer: ```suggestion
AWS Authentication utilizes a user profile specified in the user's home directory at `.aws/credentials`. Each profiles' credentials should include at minimum
```

the profile name, access key, and secret access key.

```
[profile_name]
aws_access_key_id=******
aws_secret_access_key=******
```

> This file should be placed in the same directory as the collector (within root directory).

Reviewer: Im not sure this is correct. The collector is installed at `/opt/observiq-otel-collector` on linux systems, and `C:\Program Files` on windows.

I think we can omit this line, as we already specified that the credentials live at `$HOME/.aws/credentials`.

In addition to the credentials file, there is also a `.aws/config`. This includes less sensitvie configuration options such as the region, output format, etc. A typical entry in the `config` file should look as such.

```
[profile_name]
region=us-west-2
output=json
```

More information on [AWS Configuration and Credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html)

### Environment Variables

> Alternatively, [AWS Environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html) can be specified to override a credentials file.

Reviewer: ```suggestion
Alternatively, [AWS Environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html) can be specified to override a credentials file.

You can modify the collector's environment variables by configuring a Systemd override.

Run `sudo systemctl edit observiq-otel-collector` and add your access key, secret key, and region.
```
[Service]
Environment=AWS_ACCESS_KEY_ID=******
Environment=AWS_SECRET_ACCESS_KEY=******
Environment=AWS_DEFAULT_REGION=us-west-2
```

```

Reviewer: The code block did not render correctly in my first comment.

```
[Service]
Environment=AWS_ACCESS_KEY_ID=******
Environment=AWS_SECRET_ACCESS_KEY=******
Environment=AWS_DEFAULT_REGION=us-west-2
```

## Setup

1. Once logged in, Select the `Configs` tab at the top of the Bindplane Home page.
2. Select a pre-existing config or create a new one.
3. Add a new source and select `AWS Cloudwatch`.

> 4. Once all parameters are listed, the configuration should be able to run with parameters already set to their default values.

Reviewer: I think this can be re-worded but I am not creative enough.

Reviewer 2: Also @armstrmi, does the `region` field need to match what's in the AWS config?

Author: Also @antonblock is the rewording in regards to  `4.`?

Reviewer 2: Yup, 4th point

5. Click save configuration.
6. Add Destination type of your choosing.
7. Apply Configuration to desired Agent.
8. Voila. Logs should be collecting

## Discovery Type

### Default Settings

When starting with an AWS Cloudwatch set to it's default values. One should see log collection from all logs groups with no filtering of log streams. Default polling interval for AWS Logs is set to 1 minute. So it may take a second to see any logs coming through.

### AutoDiscover

> When using Discovery Type `AutoDiscover` there are some optional parameters that can be added to reduced/ filter the amount of logs collected.

Reviewer 2: ```suggestion
When using Discovery Type `AutoDiscover` there are some optional parameters that can be added to reduced/filter the amount of logs collected.
```

- `limit`: limits the number of discovered log groups(default = 50).
- `prefix`: Prefix for log groups to limit the number of log groups discovered
  - `prefix: /aws/eks/`
  - if omitted, all log groups up to the limit will be collected from.
- `names`: A list of full log stream names to filter the discovered log groups to collect from.
  - `names: [kube-apiserver-ea9c831555adca1815ae04b87661klasdj]`
- `prefixes`: A list of log stream prefixes to filter the discovered log groups to collect from.
  - `prefixes: [kube-api-controller]`

### Named

This Discovery Type filters logs by listing only the desired log groups to collect from and omitting any other log groups. When selecting this Discovery Type, at least one log group is required otherwise no logs would be collected. When listing log groups the `ID` field of each log group instance should match the full name of the group.

Additionally, `Named` also provides `prefixes` and `names` parameters for each listed log group that filters out the listed log streams. These parameters should be listed underneath each log group's ID's as they are unique to each individual log group.

> ## Supported Platforms

Reviewer: I think this section can go to the top.

| Platform | Metrics | Logs | Traces |
| :------- | :------ | :--- | :----- |
| Linux    |         | ✓    |        |
| Windows  |         | ✓    |        |
| macOS    |         | ✓    |        |

> ## Configuration Table

Reviewer: I think this could come before default settings and auto discover sections. Perhaps those sections could be nested under this section.

| Parameter              | Type                      | Default      | Description                                                                                                 |
| :--------------------- | :------------------------ | :----------- | :---------------------------------------------------------------------------------------------------------- |
| region\*               | `enum`                    | us-east-1    | The AWS recognized region string.                                                                           |

> | profile                | `string`                  | ""           | The AWS profile used to authenticate, if none is specified the default is chosen from the list of profiles. |

Reviewer 2: Right now the default is `default`. Should we remove it, or should this be updated here?

Reviewer: Im in favor of keeping default "default". I think it improves the UI experience.

Reviewer 2: @armstrmi This should be `default`, then

the profile name, access key, and secret access key.

```
[profile_name]
aws_access_key_id=******
aws_secret_access_key=******
```

In addition to the credentials file, there is also a `.aws/config`. This includes less sensitvie configuration options such as the region, output format, etc. A typical entry in the `config` file should look as such.

```
[profile_name]
region=us-west-2
output=json
```

More information on [AWS Configuration and Credentials](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html)

### Environment Variables

Alternatively, [AWS Environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html) can be specified to override a credentials file.
You can modify the collector's environment variables by configuring a Systemd override.
Run `sudo systemctl edit observiq-otel-collector` and add your access key, secret key, and region.

```
[Service]
Environment=AWS_ACCESS_KEY_ID=******
Environment=AWS_SECRET_ACCESS_KEY=******
Environment=AWS_DEFAULT_REGION=us-west-2

> ```

Reviewer 2: I think we should also say to `sudo systemctl restart observiq-otel-collector` to be safe

Author: should we also include 
`sudo launchctl stop observiq-otel-collector`
`sudo launchctl start observiq-otel-collector`

Reviewer: > should we also include `sudo launchctl stop observiq-otel-collector` `sudo launchctl start observiq-otel-collector`

The example shown here is for Linux (Linux's Systemd service manager). We might want to include Windows before macOS.

We could just say "Restart the collector after making environment changes" instead of a platform specific command.

## Setup

1. Once logged in, Select the `Configs` tab at the top of the Bindplane Home page.
2. Select a pre-existing config or create a new one.
3. Add a new source and select `AWS Cloudwatch`.

> 4. After configuring credentials on the collector system, using the default values in the source form should enable the collector to collect logs from Cloudwatch.

Reviewer 2: ```suggestion
4. After configuring credentials using the AWS CLI on the collector system, using the default values in the source form should enable the collector to collect logs from Cloudwatch. If credentials were configured using environment variables, you will need to leave the `Profile` field blank
```

Sorry I forgot about this quirk. Is this accurate?

Author: Yes, pretty much if you intend on using env variables for authentication, the profile field should be left blank.

> ---

Reviewer: This is a good start. This source type is uniquely confusing for users because it requires additional setup. We should document the following:

- Link to AWS cli install guide (amazon docs)
- Explain that AWS cli must be configured for the `root` user (or the user running the collector. This is the root user by default for the observiq collector).
- Look into whether or not environment variables for access key and secret key can be used instead of a profile, and look into whether or not AWS CLI must be installed if this is the case.

This is probably considerable effort, and I would be okay with deferring this work to a future ticket. I just want to make sure it gets tracked because we know from experience how difficult setting this up can be for the user (from the observIQ cloud days).

Author: This is a good point, I know when demoing this, it seemed like a pretty significant point so I'd like to get @jmwilliams89 opinion on this.
